File: mote_bringup/mote_bringup/odom_tf_relay.py
"""Republish wheel odometry as an inverted TF leaf so a lidar-odometry node can
own the odom->base edge while still receiving the wheel odometry as a prior.

diff_drive publishes the wheel pose (base in odom) on a topic; this node
broadcasts its inverse as base_frame -> child_frame (a leaf). kinematic_icp,
configured with wheel_odom_frame = child_frame, reads that leaf as its motion
prior and is then free to publish the real odom -> base transform itself.
"""

import rclpy
from rclpy.node import Node
from nav_msgs.msg import Odometry
from geometry_msgs.msg import TransformStamped
from tf2_ros import TransformBroadcaster


def _rotate(q, v):
    """Rotate vector v by quaternion q = (x, y, z, w)."""
    x, y, z, w = q
    ux = 2.0 * (y * v[2] - z * v[1])
    uy = 2.0 * (z * v[0] - x * v[2])
    uz = 2.0 * (x * v[1] - y * v[0])
    return (
        v[0] + w * ux + (y * uz - z * uy),
        v[1] + w * uy + (z * ux - x * uz),
        v[2] + w * uz + (x * uy - y * ux),
    )


class OdomTfRelay(Node):
    def __init__(self):
        super().__init__("odom_tf_relay")
        self.child_frame = self.declare_parameter("child_frame", "odom_wheel").value
        self.br = TransformBroadcaster(self)
        self.create_subscription(Odometry, "odom_in", self._cb, 10)

    def _cb(self, msg):
        p = msg.pose.pose.position
        q = msg.pose.pose.orientation
        q_inv = (-q.x, -q.y, -q.z, q.w)
        ti = _rotate(q_inv, (p.x, p.y, p.z))

        t = TransformStamped()
        t.header.stamp = msg.header.stamp
        t.header.frame_id = msg.child_frame_id  # base_footprint
        t.child_frame_id = self.child_frame  # odom_wheel (leaf)
        t.transform.translation.x = -ti[0]
        t.transform.translation.y = -ti[1]
        t.transform.translation.z = -ti[2]
        t.transform.rotation.x = q_inv[0]
        t.transform.rotation.y = q_inv[1]
        t.transform.rotation.z = q_inv[2]
        t.transform.rotation.w = q_inv[3]
        self.br.sendTransform(t)


def main():
    rclpy.init()
    node = OdomTfRelay()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.try_shutdown()


if __name__ == "__main__":
    main()

File: mote_nav/src/odom_tf_relay.cpp
#include "mote_nav/odom_tf_relay.hpp"

#include <memory>
#include <string>
#include <utility>

namespace mote_nav
{

geometry_msgs::msg::TransformStamped invertOdometry(
  const nav_msgs::msg::Odometry & msg, const std::string & child_frame)
{
  const auto & p = msg.pose.pose.position;
  const auto & q = msg.pose.pose.orientation;
  const std::array<double, 4> q_inv{-q.x, -q.y, -q.z, q.w};
  const std::array<double, 3> ti = rotateVector(q_inv, {p.x, p.y, p.z});

  geometry_msgs::msg::TransformStamped t;
  t.header.stamp = msg.header.stamp;
  t.header.frame_id = msg.child_frame_id;  // base_footprint
  t.child_frame_id = child_frame;          // odom_wheel (leaf)
  t.transform.translation.x = -ti[0];
  t.transform.translation.y = -ti[1];
  t.transform.translation.z = -ti[2];
  t.transform.rotation.x = q_inv[0];
  t.transform.rotation.y = q_inv[1];
  t.transform.rotation.z = q_inv[2];
  t.transform.rotation.w = q_inv[3];
  return t;
}

OdomTfRelay::OdomTfRelay(const rclcpp::NodeOptions & options)
: rclcpp::Node("odom_tf_relay", options)
{
  child_frame_ = declare_parameter<std::string>("child_frame", "odom_wheel");
  broadcaster_ = std::make_unique<tf2_ros::TransformBroadcaster>(*this);
  subscription_ = create_subscription<nav_msgs::msg::Odometry>(
    "odom_in", rclcpp::QoS(10),
    [this](const nav_msgs::msg::Odometry::ConstSharedPtr msg) {this->onOdom(msg);});
}

void OdomTfRelay::onOdom(const nav_msgs::msg::Odometry::ConstSharedPtr msg)
{
  broadcaster_->sendTransform(invertOdometry(*msg, child_frame_));
}

}  // namespace mote_nav

#include "rclcpp_components/register_node_macro.hpp"
RCLCPP_COMPONENTS_REGISTER_NODE(mote_nav::OdomTfRelay)
